assets/js/googlesitekit/widgets/util/get-widget-class-names.js

```diff
@@ -0,0 +1,170 @@
+	if ( counter !== 9 ) {
+		return [ classNames, counter ];
+	}
+
+	classNames = [ ...classNames ];
+
+	// Start counting backwards from the last item.
+	let i = classNames.length - 1;
+
+	// Go back until counter is 0. The i >= 0 check is an extra safeguard that, with
+	// correct usage should never apply, but is still useful to avoid infinite loops
+	// if the function was used incorrectly.
+	while ( counter !== 0 && i >= 0 ) {
+		// Skip any classNames that are `null`; this happens when the component itself
+		// renders `null`.
+		if ( ! classNames[ i ] || ! Array.isArray( classNames[ i ] ) ) {
+			i--;
+			continue;
+		}
+
+		const singleWidgetClassNames = [ ...classNames[ i ] ];
+
+		// Replace the 3-column class with a 4-column class so this element goes from 1/4
+		// to 1/3 on desktop.
+		if ( singleWidgetClassNames.includes( 'mdc-layout-grid__cell--span-3-desktop' ) ) {
+			singleWidgetClassNames[ singleWidgetClassNames.indexOf( 'mdc-layout-grid__cell--span-3-desktop' ) ] = 'mdc-layout-grid__cell--span-4-desktop';
+			counter -= 3;
+		}
+
+		// Replace the 6-column class with a 8-column class so this element goes from 1/2
+		// to 2/3 on desktop.
+		if ( singleWidgetClassNames.includes( 'mdc-layout-grid__cell--span-6-desktop' ) ) {
+			singleWidgetClassNames[ singleWidgetClassNames.indexOf( 'mdc-layout-grid__cell--span-6-desktop' ) ] = 'mdc-layout-grid__cell--span-8-desktop';
+			counter -= 6;
+		}
+
+		classNames[ i ] = singleWidgetClassNames;
+		i--;
+	}
+
+	return [ classNames, counter ];
+}
+
+/**
+ * Gets widget class names for an area.
+ *
+ * @since n.e.x.t
+ *
+ * @param {Array.<Object>} activeWidgets List of active widgets.
+ * @return {Array.<string>} List of class names for active widgets.
+ */
+export function getWidgetClassNames( activeWidgets ) {
+	let counter = 0;
+	let classNames = [].fill( null, 0, activeWidgets.length );
+
+	activeWidgets.forEach( ( widget, i ) => {
+		// Let's make sure we always have an array of widths for the widget.
+		const widths = Array.isArray( widget.width ) ? widget.width : [ widget.width ];
+
+		// Get available sizes for the current widget to select the most appropriate width for the current row.
+		let sizes = widths.map( ( widgetWidth ) => [ counter + WIDTH_GRID_COUNTER_MAP[ widgetWidth ], widgetWidth ] );
+
+		// If it is the last widget, we can try to use an alternative size if it fits into the current row.
+		if ( i + 1 === activeWidgets.length ) {
+			// We need to check whether we have a size that can fit into the row and if so, try to get it.
+			const hasSizeThatCanFitIntoRow = sizes.some( ( [ widgetWidth ] ) => widgetWidth <= 12 );
+			if ( hasSizeThatCanFitIntoRow ) {
+				// Sort available sizes to have the descending order.
+				sizes = sizes.sort( ( [ a ], [ b ] ) => b - a );
+				// Filter out only those sizes that fit into the current row.
+				sizes = sizes.filter( ( [ widgetWidth ] ) => widgetWidth <= 12 );
+			}
+		}
```

Comment:
This approach doesn't address the requirement outlined in the ACs to try to fit the last widget _in a row_ into the row - the logic you have here only does something to the last widget _in the entire widget area_. Furthermore it doesn't take the current value of `counter` into account, without which I think it's impossible to address the issue.

To explain a bit more:
* The counter goes up for each widget until it reaches 12, then it goes back to 0 (meaning a new row will be started).
* If a widget increases the counter to a value greater than 12, the widget is too wide. **This is when we need to add new logic here to try to either fit this widget in or try to fill up the remaining space otherwise:**
    * If the widget has a fallback width that fits (`counter` would with that width remain <= 12), use that width. More specifically, use the first width in the array of widths for which that condition applies (if any).
    * Otherwise (the current widget "just doesn't fit"), it will for sure go into the next row (as it already would with the current logic in `develop`). In this case, check the previous widget whether it has an alternative width specified which would allow it to fill up the remaining space (i.e. increase `counter` to the maximum value <= 12).

This logic should happen whenever we reach the end of a row, and as well when we reach the last widget in the entire area.

Reply:
@felixarntz, I have updated my approach to work for every row where an alternative size can be used. So, I think we should be good now.

> Furthermore it doesn't take the current value of `counter` into account, without which I think it's impossible to address the issue.

It does take it into account and makes all calculations based on the current value of the `counter`.